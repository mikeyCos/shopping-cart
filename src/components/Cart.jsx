import PropTypes from "prop-types";
import { createContext, useContext, useState } from "react";
import { Link } from "react-router-dom";
import ProductCard from "./ProductCard";
import FormQuantity from "./FormQuantity";
import InputQuantity from "./InputQuantity";
import formatPrice from "../utilities/formatPrice";
import styles from "../styles/Cart.module.css";

const Cart = () => {
  const { cart, removeFromCart, setQuantity } = useContext(CartContext);

  const subTotal = cart.reduce((accumulator, currentCartItem) => {
    return accumulator + currentCartItem.price * currentCartItem.quantity;
  }, 0);

  return (
    <section id={styles["cart"]} role="region">
      Cart
      {cart.length === 0 && <h4>Cart is empty</h4>}
      {cart.length > 0 && (
        <section role="region" className={styles["cart-container"]}>
          {cart.map((item) => {
            return (
              <ProductCard
                key={item.id}
                className={styles["cart-product-card"]}
              >
                <ProductCard.Heading title={item.title} />
                <Link
                  to={`/shop/product/view/${encodeURIComponent(item.title)}`}
                  state={{ product: item }}
                >
                  <ProductCard.Picture
                    src={item.image}
                    alt="#"
                    className={styles["product-picture"]}
                  />
                </Link>
                <FormQuantity submitForm={() => removeFromCart(item.id)}>
                  <p>{formatPrice(item.price)}</p>
                  <div className="form-controls">
                    <InputQuantity
                      quantity={item.quantity}
                      setQuantity={(newQuantity) => {
                        console.log("previous quantity", item.quantity);
                        setQuantity(item.id, newQuantity);
                      }}
                      incrementHandler={() => {
                        const newQuantity = +item.quantity + 1;
                        newQuantity <= 999 && setQuantity(item.id, newQuantity);
                      }}
                      decrementHandler={() => {
                        const newQuantity = item.quantity - 1;
                        newQuantity >= 0 && setQuantity(item.id, newQuantity);
                      }}
                    />
                    <FormQuantity.SubmitButton text="Delete" />
                  </div>
                </FormQuantity>
              </ProductCard>
            );
          })}
          <p className={styles["subtotal"]}>
            Subtotal: <span>{formatPrice(subTotal)}</span>
          </p>
        </section>
      )}
    </section>
  );
};

/* React hook createContext
 * https://react.dev/learn/passing-data-deeply-with-context
 */
const CartContext = createContext([]);

const CartProvider = ({ children }) => {
  const [cart, setCart] = useState([]);
  /* Removing item from cart
   * Check if item is in cart
   * If item is in cart
   *  Reduce cart and add quantities of item
   * Else
   *  Add item to cart
   *
   * Update item's quantity
   */
  const addToCart = (product, quantity) => {
    const { id } = product;

    /* // Option 1:
    const newCart = productInCart
      ? cart.map((productInCart) => {
          if (productInCart.id === id)
            return {
              ...productInCart,
              quantity: productInCart.quantity + quantity,
            };
          return productInCart;
        })
      : [...cart, { ...product, quantity }];
    setCart(newCart); */

    let isInCart = false;
    const newCart = cart.map((productInCart) => {
      if (productInCart.id === id) {
        isInCart = true;
        return {
          ...productInCart,
          quantity: productInCart.quantity + quantity,
        };
      }

      return productInCart;
    });

    setCart(isInCart ? newCart : [...cart, { ...product, quantity }]);

    /* // Option 2:
    setCart((prevCart) => {
      const isInCart = prevCart.some((cartItem) => cartItem.id === product.id);

      return isInCart
        ? prevCart.reduce((accumulator, currentCartItem) => {
            return [
              ...accumulator,
              currentCartItem.id === product.id
                ? {
                    ...currentCartItem,
                    quantity: currentCartItem.quantity + quantity,
                  }
                : currentCartItem,
            ];
          }, [])
        : [...prevCart, { ...product, quantity }];
    }); */
  };

  const removeFromCart = (productID) => {
    setTimeout(() => {
      setCart((prevCart) => {
        return prevCart.filter((product) => product.id !== productID);
      });
    }, 250);
  };

  const setQuantity = (productID, quantity) => {
    if (quantity == 0) {
      removeFromCart(productID);
    } else {
      setCart((prevCart) =>
        prevCart.map((product) => {
          return product.id === productID ? { ...product, quantity } : product;
        })
      );
    }
  };

  return (
    <CartContext.Provider
      value={{ cart, addToCart, removeFromCart, setQuantity }}
    >
      {children}
    </CartContext.Provider>
  );
};

CartProvider.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};

export { Cart as default, CartContext, CartProvider };
