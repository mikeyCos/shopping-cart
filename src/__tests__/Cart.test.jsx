import { describe, expect, it, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import Cart from "../components/Cart";
import CartProvider, { CartContext } from "../context/CartContext";
import { product, products } from "./mocks";
import userEvent from "@testing-library/user-event";

describe("Cart component", () => {
  const initialEntries = [
    { pathname: "/cart", state: { previousLocation: "/shop/cart" } },
  ];

  it("'Cart is empty' is rendered", () => {
    render(
      <CartProvider>
        <MemoryRouter initialEntries={initialEntries}>
          <Cart />
        </MemoryRouter>
      </CartProvider>
    );
    const heading = screen.getByRole("heading", { level: 4 });
    expect(heading.textContent).toMatch(/Cart is empty/i);
  });

  it("Cart component matches snapshot", () => {
    const initialEntries = [
      { pathname: "/cart", state: { previousLocation: "/shop/cart" } },
    ];

    const { container } = render(
      <CartContext.Provider value={{ cart: [{ ...product, quantity: 1 }] }}>
        <MemoryRouter initialEntries={initialEntries}>
          <Cart />
        </MemoryRouter>
      </CartContext.Provider>
    );

    expect(container).matchSnapshot();
  });

  it("There are 20 products in the cart", () => {
    const initialEntries = [
      { pathname: "/cart", state: { previousLocation: "/shop/carts" } },
    ];

    const cartProducts = products.map((product) => ({
      ...product,
      quantity: Math.floor(Math.random() * 5) + 1,
    }));
    render(
      <CartContext.Provider value={{ cart: cartProducts }}>
        <MemoryRouter initialEntries={initialEntries}>
          <Cart />
        </MemoryRouter>
      </CartContext.Provider>
    );

    const productCards = screen.getAllByRole("article");
    expect(productCards.length).toBe(20);
  });

  it("The delete button's onClick handler fires and is provided with it's respective productID", async () => {
    const user = userEvent.setup();
    const initialEntries = [
      { pathname: "/cart", state: { previousLocation: "/shop/carts" } },
    ];

    const cartProducts = products.map((product) => ({
      ...product,
      quantity: Math.floor(Math.random() * 5) + 1,
    }));

    const removeFromCart = vi.fn((productID) => {});

    render(
      <CartContext.Provider
        value={{
          cart: cartProducts,
          removeFromCart,
        }}
      >
        <MemoryRouter initialEntries={initialEntries}>
          <Cart />
        </MemoryRouter>
      </CartContext.Provider>
    );

    const deleteButtons = screen.getAllByRole("button", { name: "Delete" });

    await user.click(deleteButtons[2]);
    expect(removeFromCart).toBeCalledWith(3);
  });
});
