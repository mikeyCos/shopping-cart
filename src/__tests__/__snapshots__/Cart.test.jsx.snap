// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cart component > Cart component matches snapshot 1`] = `
<section
  id="cart"
  role="region"
>
  Cart
  <article>
    <h4>
      Opna Women's Short Sleeve Moisture
    </h4>
    <picture>
      <img
        alt="#"
        loading="lazy"
        src="https://fakestoreapi.com/img/51eg55uWmdL._AC_UX679_.jpg"
        style="width: 200px;"
      />
    </picture>
    <form
      novalidate=""
    >
      <p>
        $7.95
      </p>
      <div>
        <label
          for="quantity"
        />
        <div
          class="_input-btn-container_e589d5"
        >
          <button
            class="_input-btn-left_e589d5"
            type="button"
          >
            -
          </button>
          <input
            class="quantity-input _input_e589d5"
            id="quantity"
            maxlength="3"
            minlength="1"
            name="quantity"
            pattern="/\\d{1,3}/"
            type="text"
            value="1"
          />
          <button
            class="_input-btn-right_e589d5"
            type="button"
          >
            +
          </button>
        </div>
        <p
          class="error-message"
        />
      </div>
      <button
        type="submit"
      >
        Delete
      </button>
    </form>
  </article>
  <p>
    Subtotal: 
    <span>
      $7.95
    </span>
  </p>
</section>
`;
